graphql/src/main/java/com/scalar/db/graphql/datafetcher/TransactionInstrumentation.java

```diff
@@ -0,0 +1,190 @@
+    GraphQLContext graphQLContext = executionContext.getGraphQLContext();
+
+    List<Directive> transactionDirectives =
+        executionContext
+            .getOperationDefinition()
+            .getDirectives(Constants.TRANSACTION_DIRECTIVE_NAME);
+
+    if (transactionDirectives.isEmpty()) {
+      return SimpleInstrumentationContext.noOp();
+    }
+
+    Directive directive = transactionDirectives.get(0); // @transaction is not repeatable
+    Argument txIdArg = directive.getArgument(Constants.TRANSACTION_DIRECTIVE_TX_ID_ARGUMENT_NAME);
+    Argument commitArg =
+        directive.getArgument(Constants.TRANSACTION_DIRECTIVE_COMMIT_ARGUMENT_NAME);
+
+    DistributedTransaction transaction;
+    if (txIdArg == null) {
+      // start a new transaction
+      try {
+        transaction = transactionManager.start();
+      } catch (TransactionException e) {
+        LOGGER.warn("failed to start transaction", e);
+        throw new AbortExecutionException(e);
+      }
+    } else {
+      // continue an existing transaction
+      String txId = ((StringValue) txIdArg.getValue()).getValue();
+      transaction =
+          activeTransactions
+              .get(txId)
+              .orElseThrow(
+                  () -> {
+                    String message =
+                        String.format(
+                            "The specified transaction %s is not found."
+                                + " It might have been expired",
+                            txId);
+                    LOGGER.warn(message);
+                    return new AbortExecutionException(message);
+                  });
+    }
+    graphQLContext.put(Constants.CONTEXT_TRANSACTION_KEY, transaction);
+
+    boolean isCommitTrue =
+        commitArg != null
+            && commitArg.getValue() != null
+            && ((BooleanValue) commitArg.getValue()).isValue();
+    if (isCommitTrue) {
+      return SimpleInstrumentationContext.whenCompleted(
+          (executionResult, throwable) -> {
+            activeTransactions.remove(transaction.getId());
+            try {
+              if (throwable != null) {
+                LOGGER.warn(
+                    "aborting the transaction {} since an error happened during execution",
+                    transaction.getId(),
+                    throwable);
+                transaction.abort();
+              } else {
+                try {
+                  LOGGER.info("committing the transaction {}", transaction.getId());
+                  transaction.commit();
+                } catch (CommitException | UnknownTransactionStatusException e) {
+                  LOGGER.warn(
+                      "an error happened when committing the transaction {}",
+                      transaction.getId(),
+                      e);
+                  graphQLContext.put(
+                      CONTEXT_TRANSACTION_ERROR_KEY,
+                      new ScalarDbTransactionError(
```

Comment:
It looks like it only handles commit related exceptions, correct? How about `CrudException` or `CrudConflictException`?

And one more question is how about error report for storage operations (`ExecutionException`)?

Reply:
Right, `CrudException` and `ExecutionException` are thrown from the data fetcher layer, so I'll check if they are handled properly and work on it in a separate pull request if necessary. The `TransactionInstrumentation` is only responsible for committing or aborting transactions.

Reply:
@brfrn169 I added PR #467 for reporting `CrudException` and `ExecutionException`. Please take a look! 🙇‍♂️